### PR TITLE
New: Scriplet that clean everything from a previous run 

### PR DESCRIPTION
**Pull request by Hook25**
We have quite a few lines in all templates dedicated to clean up what the previous job left behind. These are mainly used for noprovision machine. Given that I need something very similar, this is my implementation as a scriplet we can call instead of the multiple lines.

Tests:
- Missing `--im-sure` run (which I definetly intended to create and didn't do it by mistake): https://github.com/canonical/checkbox/actions/runs/11103858615/job/30846714818
- Actual run: https://github.com/canonical/checkbox/actions/runs/11104217105/job/30847831704#step:5:512

### CODE REVIEW

File: scriptlets/clean_machine [hunk 1]
```diff
@@ -0,0 +1,73 @@
+#!/bin/bash
+# this is a generic clean script that deletes any thing that may break the
```
Comment (boukeas) [inline scriptlets/clean_machine:2]: Typo "This"

File: scriptlets/clean_machine [hunk 2]
```diff
@@ -0,0 +1,73 @@
+#!/bin/bash
```
Comment (boukeas) [inline scriptlets/clean_machine:1]: This is supposed to run on the DUT, right? I think this is perhaps something that needs to be explicit in the "docstring". Also, if it's supposed to run on the device then it needs to be one of the scriptlets that is copied over to the device in the [installer script](https://github.com/canonical/hwcert-jenkins-tools/blob/fca2f9f2269dcd0f6120feaf1c75c157854accc8/install_tools.sh#L38).

Just to note (and I am really not sure of the answer myself): so far we have been running this sort of thing from the agent using individual `_run` commands. This shifts the approach to running the commands/script directly on the device. I am not sure which is best but it does need to be weighed and considered.

File: scriptlets/clean_machine [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+# delete all checkbox snaps, this includes any frontend or custom frontend
+# (they all contain checkbox in their name). This should also purge any
+# running checkbox*agent service
+CHECKBOX_SNAPS=$(snap list | \grep -E 'checkbox' | \grep -P -o "^[\w-]+")
+if [ -n "$CHECKBOX_SNAPS" ]; then
+    echo "Found the following Checkbox snaps"
+    echo "  " $CHECKBOX_SNAPS
+    echo "Deleting them..."
+    snap remove $CHECKBOX_SNAPS
```
Comment (boukeas) [inline scriptlets/clean_machine:31]: Don't you need a `sudo` here?

File: scriptlets/clean_machine [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  echo "if you are sure launch: `basename "$0"` --im-sure"
+  exit 1
+fi
+
+set +e
+
+# delete all checkbox snaps, this includes any frontend or custom frontend
+# (they all contain checkbox in their name). This should also purge any
+# running checkbox*agent service
+CHECKBOX_SNAPS=$(snap list | \grep -E 'checkbox' | \grep -P -o "^[\w-]+")
```
Comment (boukeas) [inline scriptlets/clean_machine:26]: Enlighten me: why `\grep` and not plain `grep`?

File: scriptlets/clean_machine [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+fi
+
+# delete all installed Checkbox debian packages. This should also purge any
+# previously installed provider
+CHECKBOX_DEBS=$(apt list --installed | \grep checkbox | \grep -P -o "^[\w-]+")
+if [ -n "$CHECKBOX_DEBS" ]; then
+    echo "Found the following Checkbox debs"
+    echo "  " $CHECKBOX_DEBS
+    echo "Deleting them..."
+    apt purge -y -qq $CHECKBOX_DEBS
```
Comment (boukeas) [inline scriptlets/clean_machine:41]: Don't you need a `sudo` here?

File: scriptlets/clean_machine [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+# delete any provider that was globally "developed" (installed in editable mode)
+if [ -d "/var/tmp/checkbox-providers-develop/"]; then
+  echo "Removing all developed providers"
+  rm -rf /var/tmp/checkbox-providers-develop
+fi
+
+# remove all ppa repos pre-configured as they may make the installer install
+# from the wrong source
+# leave also /ppa and /testing, legacy names of /edge and /beta
```
Comment (boukeas) [inline scriptlets/clean_machine:59]: The comment may be a little misleading here. One could it read it as "remove all these PPAs but _leave_ edge and beta" but it's obviously not what you mean.

File: scriptlets/clean_machine [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+echo "Trying to remove all Checkbox ppas"
+add-apt-repository --remove -y ppa:checkbox-dev/ppa
```
Comment (boukeas) [inline scriptlets/clean_machine:61]: Don't you need `sudo` here?

Comment (boukeas) [inline scriptlets/clean_machine:61]: I seem to remember this part being a little "noisy" in the SRU runs when the PPAs have not been previously added to the system. Would it perhaps make sense to suppress the output (e.g. redirect to `/dev/null`) and only print some sort of message when the command is successful, for example:
```
sudo add-apt-repository --remove -y ppa:checkbox-dev/$RISK > /dev/null && echo "Removed ppa:checkbox-dev/$RISK"
```
Also, would you consider using a for-loop over $RISK here?

File: scriptlets/clean_machine (hunk 5 above)
Comment (boukeas) [inline scriptlets/clean_machine:41]: Does using `apt` instead of `apt-get` still generate a warning that `apt` shouldn't be used in scripts because it's interface is not stable? If so, perhaps best to go for `apt-get`

File: scriptlets/clean_machine [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+add-apt-repository --remove -y ppa:checkbox-dev/testing
+add-apt-repository --remove -y ppa:checkbox-dev/edge
+add-apt-repository --remove -y ppa:checkbox-dev/beta
+add-apt-repository --remove -y ppa:checkbox-dev/stable
+
+# Also remove any cloned version of Checkbox in any home, as this will clash
+# with anything trying to provision from source
+CHECKBOX_SOURCE_CLONES=$(find /home/ -maxdepth 2 -name "*checkbox*" -type d)
```
Comment (boukeas) [inline scriptlets/clean_machine:69]: Just a suggestion but you can delete the directories you find directly within `find`
```
echo "Removing all clones of Checkbox on the machine"
find /home/ -maxdepth 2 -name "*checkbox*" -type d -exec rm -rf {} +
```
If I remember correctly this should also display the ones that are found and removed.

**Review by boukeas — CHANGES_REQUESTED** [review]
This is really useful and should clean up the SRU script considerably. I have made a few comments and suggestions. In the end, can you please also include a link to one or two Testflinger jobs where the scriptlet is used, as a demonstration of how it works?

File: scriptlets/clean_machine (hunk 2 above)
Comment (Hook25) [reply scriptlets/clean_machine:1]: I think this is a bit easier to read as is, as basically every  command in this would use `_run` and I can see only downsides to it. The only thing we can get out of `_run`ing every command in this via individual `_run` is network latency, or am I missing something?

File: scriptlets/clean_machine (hunk 4 above)
Comment (Hook25) [reply scriptlets/clean_machine:26]: I have a lot of aliases on my machine, I will fix this, I was copy pasting from my console

File: scriptlets/clean_machine (hunk 5 above)
Comment (Hook25) [reply scriptlets/clean_machine:41]: ditto above

File: scriptlets/clean_machine (hunk 7 above)
Comment (Hook25) [reply scriptlets/clean_machine:61]: ditto above

Comment (Hook25) [reply scriptlets/clean_machine:61]: Makes sense, originally I wanted to detect installed ppas and remove them, but that is harder than just removing them all

File: scriptlets/clean_machine (hunk 5 above)
Comment (Hook25) [reply scriptlets/clean_machine:41]: I will fix it, same issue as in `\grep`

File: scriptlets/clean_machine (hunk 8 above)
Comment (Hook25) [reply scriptlets/clean_machine:69]: yes, I wanted for "simmetry" to only print the "Removing all clones..." message when there is actually something to remove, but I can change this.

File: scriptlets/clean_machine (hunk 2 above)
Comment (Hook25) [reply scriptlets/clean_machine:1]: Also, I expected the installer script to copy everything in the scriplet dir with a glob, any reason we are maintaining an explicit list over there?

File: scriptlets/clean_machine (hunk 3 above)
Comment (boukeas) [reply scriptlets/clean_machine:31]: Yes, I don't think it's good practice to do that. Use `sudo` only when you need to.

File: scriptlets/clean_machine (hunk 7 above)
Comment (Hook25) [reply scriptlets/clean_machine:61]: I'm keeping one `apt` command, `apt list` as I don't know if there is a "clean" alternative, I think to not use it one needs to fetch the installed packages via dpkg, I would rather not

File: scriptlets/clean_machine (hunk 2 above)
Comment (Hook25) [reply scriptlets/clean_machine:1]: Fixed beside 1

File: scriptlets/clean_machine (hunk 3 above)
Comment (Hook25) [reply scriptlets/clean_machine]: Fixed

File: scriptlets/clean_machine (hunk 4 above)
Comment (Hook25) [reply scriptlets/clean_machine]: Fixed

File: scriptlets/clean_machine (hunk 5 above)
Comment (Hook25) [reply scriptlets/clean_machine]: Fixed

File: scriptlets/clean_machine (hunk 6 above)
Comment (Hook25) [reply scriptlets/clean_machine]: Fixed

File: scriptlets/clean_machine (hunk 7 above)
Comment (Hook25) [reply scriptlets/clean_machine]: Fixed

Comment (Hook25) [reply scriptlets/clean_machine]: Fixed

File: scriptlets/clean_machine (hunk 8 above)
Comment (boukeas) [reply scriptlets/clean_machine:68]: Just to clarify: you decided against using the `-exec` version?

File: install_tools.sh [hunk 9]
```diff
@@ -35,7 +35,7 @@ install_on_device() {
 
     # copy selected scriptlets over to the device
     # and then move them somewhere in the device's PATH
-    DEVICE_SCRIPTLETS=(retry check_for_packages_complete wait_for_packages_complete install_packages)
+    DEVICE_SCRIPTLETS=(retry check_for_packages_complete wait_for_packages_complete install_packages clean_machine)
```
Comment (boukeas) [inline install_tools.sh:38]: Looking at the name of the script, can I suggest renaming it to something that suggests it cleans traces of Checkbox, rather than cleaning a machine generally?

Comment (Hook25) [reply install_tools.sh:38]: I expect this script to be updated to be the one stop shop for cleaning a noprovision machine. If any command to clean up a machine has to be added, it must be added here. I only wrote the Checkbox side of business because that is what I know best

**Review by boukeas — APPROVED** [review]

File: scriptlets/clean_machine (hunk 8 above)
Comment (Hook25) [reply scriptlets/clean_machine:68]: yes as I think the current one produces a better output log

**Comment by Hook25** [comment]
Merged as workflow did pass here: https://github.com/canonical/checkbox/actions/runs/11104965148/job/30850115894

Failure is due to network fails
